### ISSUE TITLE
useAsyncQuery can't refresh

### ISSUE THREAD

**Issue by Saurou** [nuxt-community/apollo-module#496]
### Environment

I came across [different](https://github.com/nuxt-modules/apollo/issues/483) [issues](https://github.com/nuxt-modules/apollo/issues/473) [opened](https://github.com/nuxt-modules/apollo/issues/102) about what it looks to be the same problem.

I have not been able to refresh the data of useAsyncQuery in any way.

### Describe the bug

useAsyncQuery only seems to work on page refresh / landing.

### Expected behaviour

refresh() or watch makes no different (not sure they are even needed on route change).

### Additional context

Nuxt 3 documentation does not mention useAsyncQuery: is this actually usable in Nuxt 3?

**Comment by Saurou** [comment]
Looks like the issue might be related to the token.

After many tries I gave a go to [nuxt-graphql-client](https://nuxt-graphql-client.web.app/) instead. 

While testing I made the api public and everything immediately worked.
When I tried a token to access it, I experienced the same errors on pages navigation.

This time, however, I found a `retainToken` option in that documentation which, added and set to true, fixed the issue.

Not sure if apollo has a similar thing, I expect it does, though I will probably stick to `nuxt-graphql-client` at this point.

**Comment by guendev** [comment]
In Addition, The result from useAsyncQuery is unable to update automatically with Apollo Cache. I had to write a custom useAsyncQuery. It is still working for me.

```typescript
type Options<TResult, TVariables> = OptionsParameter<TResult, TVariables> & {
  key?: string
}

export const useAsyncQuery = async <TResult = any, TVariables extends OperationVariables = OperationVariables> (document: DocumentParameter<TResult, TVariables>, variables?: VariablesParameter<TVariables>, options?: Options<TResult, TVariables>) => {
  const { client } = useApolloClient()

  const fn = () => client.query<TResult, TVariables>({
    query: document as DocumentNode,
    variables: variables as TVariables
  }).then((res) => res.data)

  const key = options?.key || hash({ query: print(document as DocumentNode), variables })
  await useAsyncData(key, fn)

  return useQuery(document, variables as TVariables)
}
```
Basically, I call useAsyncData to write the result to Apollo Cache and return useQuery instead.

**Comment by Dinuz** [comment]
Indeed the refresh doesn't work at all. I am unable to refresh the data, and I see that the ref are updated, so the issue is in the refresh function.
What would be a better way to trigger a reactive action on useAsyncQuery (refreshing the data, or remaking the query), when the variables reactive change? It would be interesting if the refresh could be triggered also automatically without the need of for example clicking to trigger it.

**Comment by ssyberg** [comment]
Also experiencing this issue, I'm kinda blown away this isn't tripping everyone up. Is is just no on is using this package? What are we missing.

**Comment by mpgalaxy** [comment]
> Also experiencing this issue, I'm kinda blown away this isn't tripping everyone up. Is is just no on is using this package? What are we missing.

Same here, this is really annoying, in nuxt2 I was refreshing my queries via EventBus after closing a modal, worked perfectly fine. With this version (5.0.0-alpha.10 and before) I can call the refresh function but no refresh is happening.
useAsyncQuery works perfectly fine for fetching queries once in async setup().

**Comment by ssyberg** [comment]
We ended up doing a bunch of extra work to ensure we're only calling `useAsyncQuery` server side and `useQuery` client side. It makes our code brittle and not very DRY but it's the only thing we found could work. I wish I had a better understanding of why `useAsyncQuery` is not recommended for client side.

**Comment by mpgalaxy** [comment]
There seems to be another problem with this version. When I close my modal I get this error:
```
Uncaught TypeError: this.$apollo.destroy is not a function
    at Proxy.destroy (vue-apollo-option.esm.js?v=15a10614:1417:18)
    at chunk-UJJZ7LWV.js:4069:88
    at callWithErrorHandling (chunk-UJJZ7LWV.js:1552:18)
    at callWithAsyncErrorHandling (chunk-UJJZ7LWV.js:1560:17)
    at hook.__weh.hook.__weh (chunk-UJJZ7LWV.js:4049:19)
    at flushPostFlushCbs (chunk-UJJZ7LWV.js:1723:41)
    at render2 (chunk-UJJZ7LWV.js:7764:5)
    at Object.unmount (chunk-UJJZ7LWV.js:5225:11)
    at onClose (modal.js?v=15a10614:57:27)
    at callWithErrorHandling (chunk-UJJZ7LWV.js:1552:18)
```
This references
```
function destroy() {
  if (this.$apollo) {
    this.$apollo.destroy();
    this.$apollo = null;
  }
}
```
in vue-apollo-option.esm.js?v=15a10614.

**Comment by mpgalaxy** [comment]
**UPDATE**: @ssyberg : I just found out that [reactive queries](https://apollo.vuejs.org/api/smart-query.html) in options api are working like in nuxt2, you just have to place the apollo object **after the setup routine**, this is awesome !!! :) With that you can easily do a `this.$apollo.queries.myquery.refresh()` and everything is perfectly fine, no need for useAsyncQuery, just everything out of the box, like in nuxt2. 
```javascript
export default defineNuxtComponent({
  name: "MyTest",
  async setup(props) {
     .....
  },
  apollo: {
    setting: {
      query: gql`
        query ($name: String!) {
          setting(where: { name: $name }) {
            show
          }
        }
      `,
      variables() {
        return {
          name: "showNameOfSomeone",
        };
      },
      update: (data) => data.setting,
      error() {
        if (process.client) {
          this.$nextTick(() => {
            this.$buefy.toast.open({
              message: this.$t("loading_error"),
              type: "is-danger",
            });
          });
        }
      },
    },
   currentSchemas: {
      query: gql`
      .......
 },
 data() {
    return {
        ......
```

**Comment by rutgerbakker95** [comment]
> We ended up doing a bunch of extra work to ensure we're only calling `useAsyncQuery` server side and `useQuery` client side. It makes our code brittle and not very DRY but it's the only thing we found could work. I wish I had a better understanding of why `useAsyncQuery` is not recommended for client side.

@ssyberg Same issue here

**Comment by sneakylenny** [comment]
### Understanding the Issue with "Refresh"

#### Discovery Process
While looking into the code and experimenting, I discovered something:
The asynchronous queries are wrapped in Nuxt's [`useAsyncData`](https://nuxt.com/docs/api/composables/use-async-data), which explains why we get **`refresh`** in return instead of [vue-apollo's](https://apollo.vuejs.org/guide-composable/query.html#refetching) (the underlying package) **`refetch`**.
I actually discovered why it doesn't work when playing around with the caching options.

#### Conclusion: `refresh =\= refetch`
When executing the `refresh` method returned by `useAsyncQuery` it doesn't execute the `refetch` method from the underlying package. It just executes the function passed to the `useAsyncData` again. Which results in loading the response from cache if cache is enabled (which is enabled by default).

#### Solution/workaround

If you want to be able to refresh you'll need to globally disable the cache. Which is a bummer if you'd like to cache. You can still **enable** cache by setting it true like this: (would love to have this documented)

```ts
const { data, refresh } = await useAsyncQuery({
  query,
  variables,
  cache: true // Here
})
```

#### Caveat using this solution/workaround

For some reason when cache is globally enabled setting it false does not work due to [this line here](https://github.com/nuxt-modules/apollo/blob/64e35de7bcfc5f45cd6571841593610646868e7f/src/runtime/composables.ts#L171) which could be fixed with [this PR](https://github.com/nuxt-modules/apollo/pull/625)

### TL:DR
**Refresh** does not **"refetch"** when cache is enabled.

**Comment by germaniuz** [comment]
@timyourivh thanks for approach, it works! But cache is very important for optimization. Is it possible to fix?

**Comment by MilesWuCode** [comment]
Same issue

**Comment by LorenzoRottigni** [comment]
Hello there,
I share also my workaround, it is correctly working both for pinia store and component script setup (achieving the same desired behavior of using the official useAsyncData composable):

```
  async function search(input: SearchInput = searchInput.value): Promise<NormalizedSearchResponse | null> {
    try {
      const { data } = await useAsyncQuery<{ search: SearchResponse }>(
        SEARCH,
        { input } as SearchQueryVariables,
      )
      return data?.value?.search
        ? ({
          ...data?.value?.search,
          items: data?.value?.search?.items?.map((item: SearchResult) => normalizeSearch(item)) || []
        })
        : null
    } catch (err) {
      console.error(err)
      return null
    }
  }

  const searchResult = useAsyncData<NormalizedSearchResponse | null>(
    'products',
    () => search(searchInput.value),
    { watch: [searchInput] }
  )
```

I tried hard to use the useAsyncQuery composable and i also gave a look to the nuxt-apollo module's core... This is the only way I found to refresh a search graphql request based off the query params.

**Comment by sneakylenny** [comment]
@germaniuz :

> @timyourivh thanks for approach, it works! But cache is very important for optimization. Is it possible to fix?

I don't know if this is still relevant for you, but for caching you should consider [`pina-colada`](https://pinia-colada.esm.dev/) to fetch and cache data. Just disable the cache for this module and fetch the data using `pina-colada`, it will automatically cache the data and re-fetch when it's "stale". Take a look at [this section](https://pinia-colada.esm.dev/guide/queries.html#When-to-use-refetch-and-refresh-), they give you full control over caching and re-fetching. The beauty of this package is that it does not care how you fetch the data!

There's also [`@tanstack/vue-query`](https://tanstack.com/query/latest/docs/framework/vue/overview), which `pina-colada` is aiming to be an alternative for, but I have no experience with this package yet.
